### DISCUSSION TITLE
Using dylint with crane

### DISCUSSION THREAD

**Discussion by TheNeikos** [ipetkov/crane#672] [Ideas]
[dylint](https://github.com/trailofbits/dylint) is a neat little tool to create ones own lints. It however has some runtime features which do not jive with how I'd like things to be defined. (Aka, the nix way, reproducibly).

This is a semi-feature request, a call for help/review as well input into developing this. (Or maybe someone already made this, and I didn't find it?)

The way dylint works is:

- You run `cargo dylint`
- It loads dynamic libraries and registers their lints
- It runs those lints on your code

Luckily `dylint` is not purely runtime driven when it comes to its inputs, and allows for customizing the source of the lints:

https://github.com/trailofbits/dylint/blob/master/docs/how_dylint_works.md#how-libraries-are-found

To quote in the section "How does dylint discover lints":

> Via the DYLINT_LIBRARY_PATH environment variable. If DYLINT_LIBRARY_PATH is set when Dylint is started, Dylint treats it as a colon-separated list of paths, and searches each path for files with names of the form DLL_PREFIX LIBRARY_NAME '@' TOOLCHAIN DLL_SUFFIX (see [Library requirements](https://github.com/trailofbits/dylint/blob/master/docs/how_dylint_works.md#library-requirements) below). For each such file found, LIBRARY_NAME resolves to that file.

This should be fairly doable to have as an input, where one takes an array of derivations and renames the output dylibs to the expected format/collects them in a store folder.

----

One snag is that lints require a nightly compiler, as they need the `rustc_private` feature. I'd have to look into this if there is another way, but given that a lint does use compiler internals, 
this is probably going to be a requirement.

Is there a way to check whether the toolchain used with crane is unstable? And if so, have the user specify an unstable one explicitely or error out.

Since the lints are not part of the code, this does not influence the 'main' compilation process, but would mean that in most cases two compilers would have to be installed. One for the 'actual' compilation, and another to compile the lints/dylints.

**Comment by ipetkov** [comment]
> Is there a way to check whether the toolchain used with crane is unstable

I'm not sure if there's a good, general way to do this at evaluation time besides introspecting the derivation's version? Even if `oxalica/rust-overlay` or `fenix` set an attribute that indicates the release channel of the toolchain, there's no guarantee the caller isn't using a version from nixpkgs which might not have the attribute

If you want to do this check at build time it's probably much easier since you could run a nightly command and check for errors and bubble up a message to the user that they should be using a nightly toolchain

**Reply by TheNeikos** [reply]
Ah yes, good idea on checking in a script in a pre- phase or so!
